data-runtime/src/main/java/io/micronaut/data/runtime/intercept/criteria/FindPageSpecificationInterceptor.java

```diff
@@ -68,7 +71,18 @@ public Object intercept(RepositoryMethodKey methodKey, MethodInvocationContext<O
             count = count(methodKey, context);
         }
 
-        Page page = Page.of(resultList, getPageable(context), count);
+        Page page;
+        if (pageable.getMode() == Pageable.Mode.OFFSET) {
+            page = Page.of(resultList, pageable, count);
+        } else {
+            PreparedQuery preparedQuery = (PreparedQuery) context.getAttribute(PREPARED_QUERY_KEY).orElse(null);
+            if (preparedQuery instanceof DefaultSqlPreparedQuery<?, ?> sqlPreparedQuery) {
+                List<Pageable.Cursor> cursors = sqlPreparedQuery.createCursors(resultList, pageable);
+                page = CursoredPage.of(resultList, pageable, cursors, count);
```

Comment:
Please add a test for `Mode.CURSOR_PREVIOUS`

Looking at the code of `createCursors`
```
if (pageable.getMode() == Mode.CURSOR_PREVIOUS) {
            Collections.reverse(results);
}
```

It does mutate the order of the results

Reply:
It is covered by this section https://github.com/micronaut-projects/micronaut-data/pull/3007/files#diff-d7abd99f936af1faa577a203b9b1c985ea5c1adeeaff9a3e03b4fb52166b7161R250-R259
```
when: "The previous page is selected"
        pageable = page.previousPageable()
        page = resultFunction.apply(pageable)

        then: "it is correct"
        page.offset == 0
        page.pageNumber == 0
        page.content.size() == 10
        page.content.id == firstContent.id
        page.content.name.every{ it.startsWith("A") }
```

Reply:
It clearly looks incorrect, the order of page values shouldn't be reversed

Reply:
Looks like it is because this is being done in `DefaultSqlPreparedQuery.attachPageable` when page is cursored page
```
sort = Sort.of(orders);
                if (cursored.isBackward()) {
                    sort = reverseSort(sort);
                }
```
and then results are reversed. Maybe none of it is not needed, but @andriy-dmytruk probably had reasons why he did it this way.

Reply:
Yes, I think the collection there should be copied

Reply:
I think it's needed when going backward and if it's sorted by name for example. Then we need to get actual X number of records prior to the first one in the current cursor, so this query is executed
```
 WHERE ((person_.name < ?) OR (person_.name = ? AND person_.id < ?)) ORDER BY person_.name DESC,person_.id DESC LIMIT 10
```
to extract 10 records prior the one we need and then later reverse it to get properly sorted.

Reply:
I guess it's part of the logic